python/ray/data/_internal/split.py

```diff
@@ -0,0 +1,242 @@
+        block_rows.append(num_rows)
+    return block_rows
+
+
+def _generate_valid_indices(
+    num_rows_per_block: List[int],
+    split_indices: List[int],
+) -> List[int]:
+    """Generate valid split indices by apply min(index, total_num_rows)
+    to every index."""
+    total_rows = sum(num_rows_per_block)
+    return [min(index, total_rows) for index in split_indices]
+
+
+def _generate_per_block_split_indices(
+    num_rows_per_block: List[int],
+    split_indices: List[int],
+) -> List[List[int]]:
+    """Given num rows per block and valid split indices, generate per block split indices.
+
+    Args:
+        num_rows_per_block: num of rows per block.
+        split_indices: The (global) indices at which to split the blocks.
+    Returns:
+        Per block split indices indicates each input block's split point(s).
+    """
+    # for each split index, we iterate though the currnet input block
+    # to see if the index falls into this block. if the index
+    # falls into this block, we push it back to the current block's
+    # split indices. Otherwise, we move on to the next block.
+    per_block_split_indices = []
+    current_input_block_id = 0
+    current_block_split_indices = []
+    current_block_global_offset = 0
+    current_index_id = 0
+
+    while current_index_id < len(split_indices):
+        split_index = split_indices[current_index_id]
+        current_block_row = num_rows_per_block[current_input_block_id]
+        if split_index - current_block_global_offset <= current_block_row:
+            current_block_split_indices.append(
+                split_index - current_block_global_offset
+            )
+            current_index_id += 1
+            continue
+        per_block_split_indices.append(current_block_split_indices)
+        current_block_split_indices = []
+        current_block_global_offset += num_rows_per_block[current_input_block_id]
+        current_input_block_id += 1
+
+    # we might finished all the indices but there are still blocks left, also
+    # current_block_split_indices might not be added yet.
+    while len(per_block_split_indices) < len(num_rows_per_block):
+        per_block_split_indices.append(current_block_split_indices)
+        current_block_split_indices = []
+    return per_block_split_indices
+
+
+def _split_single_block(
+    block_id: int,
+    block: Block,
+    meta: BlockMetadata,
+    block_row: int,
+    split_indices: List[int],
+) -> Tuple[int, List[Tuple[ObjectRef[Block], BlockMetadata]]]:
+    """Split the provided block at the given indices."""
+    split_result = []
+    stats = BlockExecStats.builder()
+    block_accessor = BlockAccessor.for_block(block)
+    prev_index = 0
+    # append one more entry at the last so we don't
+    # need handle empty edge case.
+    split_indices.append(block_row)
+    for index in split_indices:
+        logger.debug(f"slicing block {prev_index}:{index}")
+        split_block = block_accessor.slice(prev_index, index, copy=True)
```

Comment:
so it could be `prev_index`==`0`, `index`==`block_row` in some edge cases, we could save a copy here. I am mostly worried about when user does an `repartition(100)` when number of blocks are already 100, which should be no copy but we end up copying every block here.